src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java

```diff
@@ -952,4 +977,62 @@ public String concreteIndex() {
             return concreteIndex;
         }
     }
+
+    /** Utility method to create either an index or a create operation depending
+     *  on the {@link OpType} of the request. */
+    private final Engine.IndexingOperation prepareIndexOperationOnPrimary(IndexRequest request, IndexShard indexShard) {
+        SourceToParse sourceToParse = SourceToParse.source(SourceToParse.Origin.PRIMARY, request.source()).type(request.type()).id(request.id())
+                .routing(request.routing()).parent(request.parent()).timestamp(request.timestamp()).ttl(request.ttl());
+        if (request.opType() == IndexRequest.OpType.INDEX) {
+            return indexShard.prepareIndex(sourceToParse, request.version(), request.versionType(), Engine.Operation.Origin.PRIMARY, request.canHaveDuplicates());
+        } else {
+            assert request.opType() == IndexRequest.OpType.CREATE : request.opType();
+            return indexShard.prepareCreate(sourceToParse,
+                    request.version(), request.versionType(), Engine.Operation.Origin.PRIMARY, request.canHaveDuplicates(), request.autoGeneratedId());
+        }
+    }
+
+    /** Execute the given {@link IndexRequest} on a primary shard, throwing a
+     *  {@link RetryOnPrimaryException} if the operation needs to be re-tried. */
+    protected final IndexResponse executeIndexRequestOnPrimary(IndexRequest request, IndexShard indexShard) throws Throwable {
```

Comment:
Sorry I'm not sure to understand what you mean? I put it there because it allowed me to share code between TransportIndexAction and TransportShardBulkAction